src/ctapipe/irf/binning.py

```diff
@@ -0,0 +1,187 @@
+"""Collection of binning related functionality for the irf tools"""
+
+import logging
+from dataclasses import dataclass
+
+import astropy.units as u
+import numpy as np
+
+from ..compat import COPY_IF_NEEDED
+from ..core import Component
+from ..core.traits import AstroQuantity, Integer
+
+__all__ = [
+    "ResultValidRange",
+    "check_bins_in_range",
+    "make_bins_per_decade",
+    "DefaultTrueEnergyBins",
+    "DefaultRecoEnergyBins",
+    "DefaultFoVOffsetBins",
+]
+
+logger = logging.getLogger(__name__)
+
+
+def check_bins_in_range(bins, valid_range, source="result", raise_error=True):
+    """
+    Check whether ``bins`` are within a ``valid_range`` and either warn
+    or raise an error if not.
+
+    Parameters
+    ----------
+    bins: u.Quantity
+        The bins to be checked.
+    valid_range: ctapipe.irf.ResultValidRange
+        Range for which bins are valid.
+        E.g. the range in which G/H cuts are calculated.
+    source: str
+        Description of which bins are being checked to give useful
+        warnings/ error messages.
+    raise_error: bool
+        Whether to raise an error (True) or give a warning (False) if
+        ``bins`` exceed ``valid_range``.
+    """
+    low = bins >= valid_range.min
+    hig = bins <= valid_range.max
+
+    if not all(low & hig):
+        with np.printoptions(edgeitems=2, threshold=6, precision=4):
+            bins = np.array2string(bins)
+            min_val = np.array2string(valid_range.min)
+            max_val = np.array2string(valid_range.max)
+            if raise_error:
+                raise ValueError(
+                    f"Valid range for {source} is {min_val} to {max_val}, got {bins}"
+                )
+            else:
+                logger.warning(
+                    f"Valid range for {source} is {min_val} to {max_val}, got {bins}",
+                )
+
+
+@u.quantity_input(e_min=u.TeV, e_max=u.TeV)
+def make_bins_per_decade(e_min, e_max, n_bins_per_decade=5):
```

Comment:
What's the reason this is here and we are not just using:
https://github.com/cta-observatory/pyirf/blob/5e8b8c9dd54d632fcb759df90a35d4029119085a/pyirf/binning.py#L100 ?

Reply:
ah, I see. This is a slightly different version that uses "roughly" n_bins_per_decade and uses exact lower, upper limits.

The pyirf version should be used for sensitivity and uses __exactly__ n_bins_per_decade.

Essentially this is linspace vs. arange

Reply:
This could be a bit clearer in the docstring